File: apps/api/presentation/restapi/availability_transformer.go
package restapi

import (
	"apps/api/domain"
	"encoding/json"
	apiContract "libs/api-contract"
	"net/http"
)

func GetAvailabilityUpdateRequest(r *http.Request) (apiContract.AvailabilityUpdateRequest, error) {
	var availabilityUpdateRequest apiContract.AvailabilityUpdateRequest
	err := json.NewDecoder(r.Body).Decode(&availabilityUpdateRequest)
	return availabilityUpdateRequest, err
}

func ToApiAvailabilityVariant(variant domain.AvailabilityVariant) apiContract.AvailabilityVariant {
	return apiContract.AvailabilityVariant{
		VariantId:         variant.VariantId,
		VariantName:       variant.VariantName,
		IsAvailable:       variant.IsAvailable,
		AvailableQuantity: ToApiQuantity(variant.AvailableQuantity),
		IsSellable:        variant.IsSellable,
		SellableQuantity:  ToApiQuantity(variant.SellableQuantity),
	}
}

func ToApiAvailabilityProduct(product domain.AvailabilityProduct) apiContract.AvailabilityProduct {
	apiVariants := []apiContract.AvailabilityVariant{}
	for _, variant := range product.Variants {
		apiVariants = append(apiVariants, ToApiAvailabilityVariant(variant))
	}

	return apiContract.AvailabilityProduct{
		ProductId:            product.ProductId,
		ProductName:          product.ProductName,
		CategoryId:           product.CategoryId,
		CategoryName:         product.CategoryName,
		AvailabilityTracking: string(product.AvailabilityTracking),
		IsAvailable:          product.IsAvailable,
		AvailableQuantity:    ToApiQuantity(product.AvailableQuantity),
		IsSellable:           product.IsSellable,
		SellableQuantity:     ToApiQuantity(product.SellableQuantity),
		Variants:             apiVariants,
	}
}

func ToDomainQuantity(quantity *int64) *int {
	if quantity == nil {
		return nil
	}
	value := int(*quantity)
	return &value
}

func ToAvailabilityProductUpdates(requests []apiContract.AvailabilityProductUpdateRequest) []domain.AvailabilityProductUpdate {
	updates := make([]domain.AvailabilityProductUpdate, 0, len(requests))
	for _, request := range requests {
		updates = append(updates, domain.AvailabilityProductUpdate{
			ProductId:         request.ProductId,
			IsAvailable:       request.IsAvailable,
			AvailableQuantity: ToDomainQuantity(request.AvailableQuantity),
		})
	}
	return updates
}

func ToAvailabilityVariantUpdates(requests []apiContract.AvailabilityVariantUpdateRequest) []domain.AvailabilityVariantUpdate {
	updates := make([]domain.AvailabilityVariantUpdate, 0, len(requests))
	for _, request := range requests {
		updates = append(updates, domain.AvailabilityVariantUpdate{
			VariantId:         request.VariantId,
			IsAvailable:       request.IsAvailable,
			AvailableQuantity: ToDomainQuantity(request.AvailableQuantity),
		})
	}
	return updates
}

File: libs/ui/src/data/api/availability.transformer.ts
// eslint-disable-next-line @nx/enforce-module-boundaries
import {
  AvailabilityMovement as ApiAvailabilityMovement,
  AvailabilityProduct as ApiAvailabilityProduct,
  AvailabilityVariant as ApiAvailabilityVariant,
} from '../../../../api-contract/src';
import {
  AvailabilityForm,
  AvailabilityMovement,
  AvailabilityProduct,
  AvailabilityVariant,
} from '../../domain';

export function toAvailabilityVariant(
  variant: ApiAvailabilityVariant
): AvailabilityVariant {
  return {
    variantId: variant.variantId,
    variantName: variant.variantName,
    isAvailable: variant.isAvailable,
    availableQuantity: variant.availableQuantity,
    isSellable: variant.isSellable,
    sellableQuantity: variant.sellableQuantity,
  };
}

export function toAvailabilityProduct(
  product: ApiAvailabilityProduct
): AvailabilityProduct {
  return {
    productId: product.productId,
    productName: product.productName,
    categoryId: product.categoryId,
    categoryName: product.categoryName,
    availabilityTracking: product.availabilityTracking,
    isAvailable: product.isAvailable,
    availableQuantity: product.availableQuantity,
    isSellable: product.isSellable,
    sellableQuantity: product.sellableQuantity,
    variants: product.variants.map(toAvailabilityVariant),
  };
}

export function toAvailabilityMovement(
  movement: ApiAvailabilityMovement
): AvailabilityMovement {
  return {
    id: movement.id,
    productId: movement.productId,
    variantId: movement.variantId,
    delta: movement.delta,
    resultingQuantity: movement.resultingQuantity,
    reason: movement.reason,
    transactionId: movement.transactionId,
    note: movement.note,
    createdAt: movement.createdAt,
  };
}

export function toApiAvailabilityUpdateRequest(form: AvailabilityForm) {
  return {
    products: form.products.map(({ productId, isAvailable, availableQuantity }) => ({
      productId,
      isAvailable,
      availableQuantity,
    })),
    variants: form.variants.map(({ variantId, isAvailable, availableQuantity }) => ({
      variantId,
      isAvailable,
      availableQuantity,
    })),
  };
}
